File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionCeil.java
package org.hisp.dhis.rules.functions;

import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
import org.hisp.dhis.rules.parser.expression.function.ScalarFunctionToEvaluate;

import static org.apache.commons.lang3.math.NumberUtils.toDouble;
import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext;
import static org.hisp.dhis.antlr.AntlrParserUtils.castDouble;

public class RuleFunctionCeil
    extends ScalarFunctionToEvaluate
{
    @Override
    public Object evaluate( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        return String.valueOf( (long) Math.ceil( toDouble( visitor.castStringVisit( ctx.expr( 0 ) ), 0.0 ) ) );
    }

    @Override
    public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        castDouble( visitor.visit( ctx.expr( 0 ) ) );
        return CommonExpressionVisitor.DEFAULT_DOUBLE_VALUE;
    }
}

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionCeil.kt
package org.hisp.dhis.rules.functions

import org.hisp.dhis.rules.RuleVariableValue
import org.hisp.dhis.rules.toDouble
import kotlin.math.ceil

class RuleFunctionCeil : RuleFunction() {

    override fun evaluate(arguments: List<String?>,
                          valueMap: Map<String, RuleVariableValue>?, supplementaryData: Map<String, List<String>>?): String {
        when {
            arguments.size != 1 -> throw IllegalArgumentException("One argument was expected, ${arguments.size} were supplied")
            else -> return ceil(arguments[0].toDouble(0.0)).toLong().toString()
        }
    }

    companion object {
        const val D2_CEIL = "d2:ceil"

        @JvmStatic
        fun create() = RuleFunctionCeil()
    }
}
